### PR TITLE
Call `gc.collect` after each test only if the user asks for it

### LINKED ISSUE #6705: The test suite spends 20% of it's time in `gc.collect()`
```
Thu Jul 07 14:33:31 2016    matplotlib_cProfile_fulltest.pstat

         192634679 function calls (182051956 primitive calls) in 345.231 seconds

   Ordered by: internal time
   List reduced from 9374 to 3 due to restriction <3>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     2986   75.105    0.025   75.107    0.025 {gc.collect}
     1250   19.139    0.015   19.139    0.015 {matplotlib._png.write_png}
     2257    9.801    0.004    9.801    0.004 {matplotlib._png.read_png_int}
```

Mostly this is due to [this call](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/testing/decorators.py#L75)

Python is very memory effective because of reference counting (object is freed when refcount drops to zero even with `gc.disable`) and `gc.collect` only forces him to run cyclic garbage collector immediately. If it is not acceptable to to remove this call, can we have option to tuning this?

### PR DESCRIPTION
**Pull request by Kojoley**
Fixes #6705

### CODE REVIEW

**Comment by WeatherGod** [comment]
I have restarted the py2.7 test two times, and it has always failed in the same spot. The other tests are also taking longer to run. I wonder if the memory usage isn't staying low and is wasting time swap thrashing?

**Comment by WeatherGod** [comment]
The particular failure is one of those weird ones that we have been noticing (image size mismatch for the pdf comparison for the hatching test). Perhaps the gc.collect() was masking a deeper problem?

**Comment by Kojoley** [comment]
I've tried to run this test locally, but got this:
```
>python tests.py matplotlib.tests.test_backend_pdf:test_hatching_legend
S
----------------------------------------------------------------------
Ran 1 test in 0.051s

OK (SKIP=1)
```
Seems that I'm missing something to perform `pdf` tests...

File: lib/matplotlib/testing/enforcedgc.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import os
+from nose.plugins import Plugin
+
+
+class EnforcedGC(Plugin):
+    """This plugin adds option to call ``gc.collect`` after each test"""
+    enabled = False
+
+    def options(self, parser, env=os.environ):
+        env_opt = 'COLLECT_GC_AFTER_TEST'
```
Comment (QuLogic) [inline lib/matplotlib/testing/enforcedgc.py]: The name is a bit weird; you don't collect the GC, the GC does the collecting. Maybe just drop collect from the name.

Comment (Kojoley) [inline lib/matplotlib/testing/enforcedgc.py]: It was `CALL_GC_COLLECT_AFTER_TEST` and then I have reduced it.
I can't name things. T_T

**Comment by jenshnielsen** [comment]
The failure was caused by an image in test_axes leaking into the test_backendpdf tests which runs right after the axes tests. See #6720 for the fix. It seems like the garbage collect mostly ensured that the figure  was destroyed before the next test masking the missing cleanup decorator

**Comment by jenshnielsen** [comment]
#6113 which is the same fix as #6120 has now been merged to master. If you rebase on top of current master this will hopefully start working as expected

**Comment by Kojoley** [comment]
All tests have passed after rebasing. :+1:

**Comment by jenshnielsen** [comment]
I'm 👍  on merging this now. It seems that the speedup is most significant on the single core tests (python 3.5 and OSX) but that is certainly important too

**Pull request merged**